File: 705.Design-HashSet.py
# https://leetcode.com/problems/design-hashset/description/
#
# algorithms
# Easy (37.78%)
# Total Accepted:    5.1k
# Total Submissions: 13.5k
# beats 93.32% of python submissions


class MyHashSet(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.hash_map = {}

    def add(self, key):
        """
        :type key: int
        :rtype: void
        """
        self.hash_map[key] = self.hash_map.get(key, 0) + 1

    def remove(self, key):
        """
        :type key: int
        :rtype: void
        """
        if self.contains(key):
            del self.hash_map[key]

    def contains(self, key):
        """
        Returns true if this set contains the specified element
        :type key: int
        :rtype: bool
        """
        return key in self.hash_map

File: 705.Design-HashSet.java
// https://leetcode.com/problems/design-hashset/
//
// algorithms
// Easy (57.08%)
// Total Accepted: 32,658
// Total Submissions: 57,210
// beats 100.0% of java submissions
// 用位图做更好，或者直接自己搞个位图，用 byte 数组

class MyHashSet {

    private boolean[] hashSet;

    /** Initialize your data structure here. */
    public MyHashSet() {
        this.hashSet = new boolean[1000001];
    }

    public void add(int key) {
        this.hashSet[key] = true;
    }

    public void remove(int key) {
        this.hashSet[key] = false;
    }

    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        return hashSet[key];
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such: MyHashSet obj
 * = new MyHashSet(); obj.add(key); obj.remove(key); boolean param_3 =
 * obj.contains(key);
 */